### PR TITLE
THREESCALE-9596: Upgrade to Sidekiq 7

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

Now we are going to have a RHEL8 based image soon it's time to work on upgrading to Sidekiq 7 again. This is required to add support for TLS and ACL for Redis: https://github.com/3scale/porta/pull/3572.

This PR is basically a reversion of https://github.com/3scale/porta/pull/3608 with a few changes to adapt it to the last alpha. However, one of the new changes is replacing `include Sidekiq::Worker` by `include Sidekiq::Job`. Most of changes in this PR are about that.

**Which issue(s) this PR fixes** 

[THREESCALE-9596](https://issues.redhat.com/browse/THREESCALE-9596)
[THREESCALE-10207](https://issues.redhat.com/browse/THREESCALE-10207)

**Verification steps** 

Sidekiq should work without issues. One particular scenario to consider is adding jobs with Sidekiq 6 and perform them with Sidekiq 7, those jobs should be performed successfully.

**Special notes for your reviewer**:

- Problem with `sidekiq-throttled`:
  - Currently we are using `0.15.0`. However this version doesn't support `ActiveJob`, which means that throttling and concurrency control just never worked for us, and this kind of calls are in practice being ignored: https://github.com/3scale/porta/blob/bc9f1d59a3b850c791fef40191c06b3592f2ea74/app/workers/backend_metric_worker.rb#L8-L14
  - Support for `ActiveJob` was partially added in https://github.com/ixti/sidekiq-throttled/pull/116 and https://github.com/ixti/sidekiq-throttled/pull/151
  - Last version `1.3.0` includes above fixes, but still doesn't support `ActiveJob` with arguments. I opened a PR for that: https://github.com/ixti/sidekiq-throttled/pull/184
  - While my PR gets reviewed and merged, you can point to my fork in order to have a fully functional Sidekiq 7 in case you want to make tests locally:
    ```ruby
    gem 'sidekiq-throttled', '~> 1.3.0', github: 'jlledom/sidekiq-throttled', branch: 'activejob-args'
    ```
- I'm forcing to use `redis-client (= 0.20.0)` because of https://github.com/redis-rb/redis-client/issues/182

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -54,10 +54,10 @@ gem '3scale_time_range', '0.0.6'
 gem 'statsd-ruby', require: false
 
 # Sidekiq
-gem 'sidekiq', '~> 6.4.0', require: %w[sidekiq sidekiq/web]
-gem 'sidekiq-batch'
+gem 'sidekiq', '~> 7', require: %w[sidekiq sidekiq/web]
+gem 'sidekiq-batch', github: '3scale/sidekiq-batch', branch: 'redis-client'
 gem 'sidekiq-cron', require: %w[sidekiq/cron sidekiq/cron/web]
-gem 'sidekiq-throttled'
+gem 'sidekiq-throttled', '~> 1.4.0'
```
Comment (akostadinov) [inline Gemfile:60]: Does this support ActiveJob fully now?

File: Gemfile.lock [hunk 2]
```diff
@@ -28,6 +28,14 @@ GIT
       nokogiri (~> 1)
       rspec (>= 3.0.0.a, < 4)
 
+GIT
+  remote: https://github.com/3scale/sidekiq-batch.git
+  revision: f1c56d0b8445fd3eb337614aedba1752aa19938e
+  branch: redis-client
```
Comment (akostadinov) [inline Gemfile.lock:34]: Does this have a native component? It will requre some sort of a fix for prod images.

File: Gemfile [hunk 3]
```diff
@@ -54,10 +54,10 @@ gem '3scale_time_range', '0.0.6'
 gem 'statsd-ruby', require: false
 
 # Sidekiq
-gem 'sidekiq', '~> 6.4.0', require: %w[sidekiq sidekiq/web]
-gem 'sidekiq-batch'
+gem 'sidekiq', '~> 7', require: %w[sidekiq sidekiq/web]
+gem 'sidekiq-batch', github: '3scale/sidekiq-batch', branch: 'redis-client'
```
Comment (akostadinov) [inline Gemfile:58]: this one, maybe add some code comment as to why we need a custom version so that later it removes from the guesswork when we try to upgrade

File: Gemfile [hunk 4]
```diff
@@ -106,7 +106,7 @@ gem 'ratelimit'
 gem 'recaptcha', '~> 5.16.0'
 gem 'redcarpet', '~>3.5.1', require: false
 gem 'RedCloth', '~>4.3', require: false
-gem 'redis', '~> 4.2.0', require: ['redis', 'redis/connection/hiredis']
+gem 'redis', require: ['redis']
```
Comment (akostadinov) [inline Gemfile:109]:
```suggestion
gem 'redis'
```

File: app/lib/backend/storage.rb [hunk 5]
```diff
@@ -7,8 +7,7 @@ def self.parse_config
       config = File.read("#{Rails.root}/config/backend_redis.yml")
       config = ERB.new(config).result(binding)
       config = YAML.load(config)
-      config = config.fetch(Rails.env).symbolize_keys
-      config.reverse_merge(logger: Rails.logger)
+      config.fetch(Rails.env).deep_symbolize_keys
```
Comment (akostadinov) [inline app/lib/backend/storage.rb:10]: why removed the `reverse_merge`?

File: app/workers/zync_worker.rb [hunk 6]
```diff
@@ -139,7 +139,7 @@ def on_complete(_bid, options)
     # The number of retries is usually controlled at the origin of the failure using ThreeScale::SidekiqRetrySupport::Worker#last_attempt?,
     # but in this case we are not really using Sidekiq retries, but rather re-enqueueing the job manually here
     manual_retry_count = options['manual_retry_count'].to_i + 1
-    perform_async(event.event_id, event.data.as_json, manual_retry_count)
+    perform_async(event.event_id, event.data.to_json, manual_retry_count)
```
Comment (akostadinov) [inline app/workers/zync_worker.rb:142]: why is this needed? I thought simple json compatible structures should be supported

File: config/initializers/redis_hacks.rb
Comment (akostadinov) [inline config/initializers/redis_hacks.rb:1]: did you test this or just too old with a questionable usability?

File: openshift/system/config/redis.yml
Comment (akostadinov) [inline openshift/system/config/redis.yml:1]: Ideally this file should be removed and only the file in `config/` would remain. See https://issues.redhat.com/browse/THREESCALE-10924

File: openshift/system/config/backend_redis.yml
Comment (akostadinov) [inline openshift/system/config/backend_redis.yml:1]: Ideally this file should be removed and only the file in `config/` would remain. See https://issues.redhat.com/browse/THREESCALE-10924

File: config/initializers/sidekiq.rb [hunk 7]
```diff
@@ -13,7 +10,6 @@
     config.try(:reliable!)
 
     config.redis = ThreeScale::RedisConfig.new(System::Application.config.sidekiq).config
-    config.error_handlers << System::ErrorReporting.method(:report_error)
```
Comment (akostadinov) [inline config/initializers/sidekiq.rb:16]: removing this, how do we ensure any errors are both - logged and reported to bugsnag?

File: Gemfile (hunk 1 above)
Comment (jlledom) [reply Gemfile:60]: Yes, at least all scenarios I found locally. I wouldn't be surprised if we find new failing scenarios in staging or production.

File: Gemfile.lock (hunk 2 above)
Comment (jlledom) [reply Gemfile.lock:34]: No native component, and they have progressed on supporting Sidekiq 7 but still haven't launched a release. I hope they do it soon so we can get rid of the fork.

File: Gemfile (hunk 3 above)
Comment (jlledom) [reply Gemfile:58]: I'll do

File: app/lib/backend/storage.rb (hunk 5 above)
Comment (jlledom) [reply app/lib/backend/storage.rb:10]: In fact, tests fail.

Comment (jlledom) [reply app/lib/backend/storage.rb:10]: Confirmed, the `:logger` param breaks the redis client when porta tries to access to backend DB directly, for instance, to get the amount of `hits` for an application. If you add `config.reverse_merge(logger: Rails.logger)` back and try to open an App screen, you'll see the error in the console when it tries to retrieve the usage data.

File: openshift/system/config/redis.yml
Comment (jlledom) [reply openshift/system/config/redis.yml:1]: I'm all for removing thing.

File: config/initializers/sidekiq.rb (hunk 7 above)
Comment (jlledom) [reply config/initializers/sidekiq.rb:16]: This was required to report to Bugsnag, but the last Bugsnag release adds support for Sidekiq 7 and the integration is transparent, so we don't need this hack anymore.

File: app/lib/backend/storage.rb (hunk 5 above)
Comment (akostadinov) [reply app/lib/backend/storage.rb:10]: So can we in some other way set logger? Or is this not needed anymore?

**Comment by akostadinov** [comment]
In this branch I still see 
```
🐚 rg Sidekiq::Worker
features/support/initial_data.rb
4:  Sidekiq::Worker.clear_all

features/support/sphinx.rb
9:  Sidekiq::Worker.clear_all

test/test_helpers/sidekiq.rb
14:        ::Sidekiq::Worker.clear_all
18:        ::Sidekiq::Worker.clear_all
23:      ::Sidekiq::Worker.drain_all

features/support/email.rb
18:  Sidekiq::Worker.clear_all

features/step_definitions/background_jobs_steps.rb
4:  Sidekiq::Worker.clear_all
```

File: config/initializers/sidekiq.rb [hunk 8]
```diff
@@ -2,9 +2,6 @@
 
 require 'three_scale/sidekiq_retry_support'
 require 'three_scale/sidekiq_logging_middleware'
-require 'sidekiq/throttled'
-
-Sidekiq::Throttled.setup!
```
Comment (mayorova) [inline config/initializers/sidekiq.rb:7]: Why don't we need this anymore?...

File: config/initializers/sidekiq.rb (hunk 7 above)
Comment (mayorova) [reply config/initializers/sidekiq.rb:16]: And are we still logging the errors? `report_error` also logs them.

**Comment by mayorova** [comment]
> In this branch I still see
> 
> ```
> 🐚 rg Sidekiq::Worker
> features/support/initial_data.rb
> 4:  Sidekiq::Worker.clear_all
> 
> features/support/sphinx.rb
> 9:  Sidekiq::Worker.clear_all
> 
> test/test_helpers/sidekiq.rb
> 14:        ::Sidekiq::Worker.clear_all
> 18:        ::Sidekiq::Worker.clear_all
> 23:      ::Sidekiq::Worker.drain_all
> 
> features/support/email.rb
> 18:  Sidekiq::Worker.clear_all
> 
> features/step_definitions/background_jobs_steps.rb
> 4:  Sidekiq::Worker.clear_all
> ```

Good point. Also, just a suggestion - the `::Job` is already available in our current version. Maybe we could all the renaming (and ONLY the renaming) in a separate PR, and this way reduce the scope of the upgrade?

File: app/lib/backend/storage.rb (hunk 5 above)
Comment (jlledom) [reply app/lib/backend/storage.rb:10]: `redis-rb` doesn't provide a way to set a logger after version 5.0. Check this: https://github.com/redis/redis-rb/issues/1182

**Comment by jlledom** [comment]
> Good point. Also, just a suggestion - the `::Job` is already available in our current version. Maybe we could all the renaming (and ONLY the renaming) in a separate PR, and this way reduce the scope of the upgrade?

It's a good idea but I think it's late now, because I already pushed the changes along with other changes in the same commit, so I would need to revert them manually in order to create the other PR. I think it's not worth the trouble. However if you think it's important I'll do it.

File: app/lib/backend/storage.rb (hunk 5 above)
Comment (akostadinov) [reply app/lib/backend/storage.rb:10]: So basically there is no logging now. Cool. No feature, no problems.

File: config/initializers/sidekiq.rb (hunk 7 above)
Comment (jlledom) [reply config/initializers/sidekiq.rb:16]: I've been checking this: It logs in the console and report to Bugsnag, but the logs in the console are not different:

Before:
```
Exception -- {:exception=>{:class=>NoMethodError, :message=>"undefined method `no_method' for nil:NilClass", :backtrace=>["/home/jlledom/redhat/porta/app/workers/sphinx_account_indexation_worker.rb:12:in `perform'", "/home/jlledom/.asdf/installs/ruby/2.7.8/lib/ruby/gems/2.7.0/gems/activejob-6.0.6.1/lib/active_job/execution.rb:40:in `block in perform_now'", "/home/jlledom/.asdf/installs/ruby/2.7.8/lib/ruby/gems/2.7.0/gems/activesupport-6.0.6.1/lib/active_support/callbacks.rb:112:in `block in run_callbacks'", "/home/jlledom/.asdf/installs/ruby/2.7.8/lib/ruby/gems/2.7.0/gems/bugsnag-6.26.0/lib/bugsnag/integrations/rails/active_job.rb:38:in `block (2 levels) in included'"]}, :parameters=>{:context=>"Job raised exception", :job=>{"retry"=>true, "queue"=>"default", "class"=>"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper", "wrapped"=>"SphinxAccountIndexationWorker", "args"=>[{"job_class"=>"SphinxAccountIndexationWorker", "job_id"=>"e1e4e6fd-28c9-4b52-83b9-be32b6c89c8c", "provider_job_id"=>nil, "queue_name"=>"default", "priority"=>nil, "arguments"=>[{"_aj_serialized"=>"ActiveJob::Serializers::ModuleSerializer", "value"=>"Account"}, 3], "executions"=>0, "exception_executions"=>{}, "locale"=>"en", "timezone"=>"UTC", "enqueued_at"=>"2024-04-15T11:47:04Z"}], "jid"=>"6237b55784171ae872bad35d", "created_at"=>1713181624.713827, "enqueued_at"=>1713181927.5125213, "error_message"=>"undefined method `no_method' for nil:NilClass", "error_class"=>"NoMethodError", "failed_at"=>1713181624.7322717, "retry_count"=>3, "retried_at"=>1713181886.8906841}, :jobstr=>"{\"retry\":true,\"queue\":\"default\",\"class\":\"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper\",\"wrapped\":\"SphinxAccountIndexationWorker\",\"args\":[{\"job_class\":\"SphinxAccountIndexationWorker\",\"job_id\":\"e1e4e6fd-28c9-4b52-83b9-be32b6c89c8c\",\"provider_job_id\":null,\"queue_name\":\"default\",\"priority\":null,\"arguments\":[{\"_aj_serialized\":\"ActiveJob::Serializers::ModuleSerializer\",\"value\":\"Account\"},3],\"executions\":0,\"exception_executions\":{},\"locale\":\"en\",\"timezone\":\"UTC\",\"enqueued_at\":\"2024-04-15T11:47:04Z\"}],\"jid\":\"6237b55784171ae872bad35d\",\"created_at\":1713181624.713827,\"enqueued_at\":1713181927.5125213,\"error_message\":\"undefined method `no_method' for nil:NilClass\",\"error_class\":\"NoMethodError\",\"failed_at\":1713181624.7322717,\"retry_count\":3,\"retried_at\":1713181886.8906841}"}}
```

Now:
```
WARN: {"context":"Job raised exception","job":{"retry":true,"queue":"default","class":"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper","wrapped":"SphinxAccountIndexationWorker","args":[{"job_class":"SphinxAccountIndexationWorker","job_id":"e1e4e6fd-28c9-4b52-83b9-be32b6c89c8c","provider_job_id":null,"queue_name":"default","priority":null,"arguments":[{"_aj_serialized":"ActiveJob::Serializers::ModuleSerializer","value":"Account"},3],"executions":0,"exception_executions":{},"locale":"en","timezone":"UTC","enqueued_at":"2024-04-15T11:47:04Z"}],"jid":"6237b55784171ae872bad35d","created_at":1713181624.713827,"enqueued_at":1713183932.1412907,"error_message":"undefined method `no_method' for nil:NilClass","error_class":"NoMethodError","failed_at":1713181624.7322717,"retry_count":4,"retried_at":1713183679.3011074}}
```

Reports to Bugsnag are also different:

Before: https://app.bugsnag.com/3scale-networks-sl/system/errors/6617c2a7d6a384000858a91e?filters[error.status]=open&filters[event.since]=30d&event_id=661d14e800e6f51f0b950000

Now: https://app.bugsnag.com/3scale-networks-sl/system/errors/6617c2a7d6a384000858a91e?filters[error.status]=open&filters[event.since]=30d&event_id=661d180e00e6d9d89f420000

They are more detailed now, though they were marked as `handled` before.

I checked I could just return to the old system and change nothing. Opinions?

Comment (akostadinov) [reply config/initializers/sidekiq.rb:16]: Nice!

Comment (akostadinov) [reply config/initializers/sidekiq.rb:16]: New is fine for me.

Comment (mayorova) [reply config/initializers/sidekiq.rb:16]: Yeah, that "sidekiq" tab (which I guess is added automatically) looks good.
Thanks for checking @jlledom !

File: config/initializers/sidekiq.rb (hunk 8 above)
Comment (mayorova) [reply config/initializers/sidekiq.rb:7]: Thanks for the explanation @jlledom

**Review by mayorova — DISMISSED** [review]

**Comment by jlledom** [comment]
@akostadinov @mayorova I merged from master and now I lost the approvals. Could you approve again, please? 🙏

**Review by mayorova — APPROVED** [review]

File: config/initializers/sidekiq.rb (hunk 8 above)
Comment (akostadinov) [reply config/initializers/sidekiq.rb:7]: But in the [readme](https://github.com/ixti/sidekiq-throttled/tree/main) I still see `require "sidekiq/throttled"` is needed :thinking:

Comment (jlledom) [reply config/initializers/sidekiq.rb:7]: @akostadinov We don't need to `require 'sidekiq/throttled'` because it's already being required from:

https://github.com/3scale/porta/blob/89a91607b8387892967a0b9315cca9e0d36c99e8/config/application.rb#L38
